src/map.cpp

```diff
@@ -5083,15 +5083,11 @@ std::list<item> map::use_charges(const tripoint &origin, const int range,
         // if vehicle has FAUCET can use contents from any of the tanks
         if( v->has_part( "FAUCET" ) ) {
             for( auto &pt : v->parts ) {
-                if( pt.is_tank() ) {
-                    for( const auto &obj : pt.contents() ) {
-                        if( obj.typeId() == type ) {
-                            ret.push_back( pt.drain( quantity ) );
-                            quantity -= ret.back().charges;
-                            if( quantity == 0 ) {
-                                return ret;
-                            }
-                        }
+                if( pt.is_tank() && pt.ammo_current() == type ) {
```

Comment:
> Also, if ammo_current() returned a matching type for something that is not a tank, this could result in breakage. The intent is to craft from tanks only.

`drain()` should exclude everything that you can't drain from (and it does). `is_tank()` is IMO too "specific" here. Imagine that we introduced multiple types of water sources for vehicles (say, giant funnels, windtraps, e.t.c.). You would need to add the checks everywhere to be able to drain them: `if( is_tank() || is_funnel() || is_windtrap() || ... ) { drain() }`.

The point is: the only thing you need to know about the part here is whether you can drain it for your purposes.

>Should I add a comment to this effect?

I still think that the first check should be dropped.